db/impl/consul/client.go

```diff
@@ -0,0 +1,274 @@
+		_, err := c.client.KV().Delete(c.qualified(path), nil)
+		return err
+	})
+}
+
+// Watch watches stuff.
+func (c *Client) Watch(obj db.Entity) (chan db.Entity, chan error) {
+	path, err := obj.Path()
+	if err != nil {
+		errChan := make(chan error, 1)
+		errChan <- err
+		return nil, errChan
+	}
+
+	// these paths are qualified in watchInternal
+	return helpers.WrapWatch(c, obj, path, false, c.watchers, &c.watcherMutex, c.watchInternal)
+}
+
+func (c *Client) watchInternal(wi helpers.WatchInfo) {
+	var wp *watch.WatchPlan
+	var err error
+
+	if wi.Recursive {
+		wp, err = watch.Parse(map[string]interface{}{
+			"type":   "keyprefix",
+			"prefix": c.qualified(wi.Path),
+		})
+	} else {
+		wp, err = watch.Parse(map[string]interface{}{
+			"type": "key",
+			"key":  c.qualified(wi.Path),
+		})
+	}
+
+	if err != nil {
+		wi.ErrorChan <- err
+		return
+	}
+
+	wp.Handler = func(u uint64, i interface{}) {
+		if i == nil {
+			return
+		}
+
+		switch i.(type) {
+		case api.KVPairs:
+			for _, pair := range i.(api.KVPairs) {
+				if pair == nil {
+					continue
+				}
+
+				this, err := helpers.ReadAndSet(c, wi.Object, pair.Key, pair.Value)
+				if err != nil {
+					wi.ErrorChan <- err
+					continue
+				}
+
+				wi.ReturnChan <- this
+			}
+		case *api.KVPair:
+			this, err := helpers.ReadAndSet(c, wi.Object, i.(*api.KVPair).Key, i.(*api.KVPair).Value)
+			if err != nil {
+				wi.ErrorChan <- err
+				return
+			}
+
+			wi.ReturnChan <- this
+		default:
+			logrus.Errorf("received invalid pair %+v during watch", i)
+		}
+	}
+
+	go func(wp *watch.WatchPlan) {
+		if err := wp.Run(c.config.Address); err != nil {
+			wi.ErrorChan <- err
+		}
+	}(wp)
+
+	go func(wp *watch.WatchPlan, wi helpers.WatchInfo) {
+		<-wi.StopChan
+		(*wp).Stop()
+	}(wp, wi)
+}
+
+// WatchPrefix watches a directory prefix for changes.
+func (c *Client) WatchPrefix(obj db.Entity) (chan db.Entity, chan error) {
+	// these paths are qualified in watchInternal
+	return helpers.WrapWatch(c, obj, obj.Prefix(), true, c.watchers, &c.watcherMutex, c.watchInternal)
+}
+
+// WatchStop stops a watch.
+func (c *Client) WatchStop(obj db.Entity) error {
+	path, err := obj.Path()
+	if err != nil {
+		return err
+	}
+
+	return helpers.WatchStop(c, path, c.watchers, &c.watcherMutex)
+}
+
+// List takes an Entity which it will then populate an []Entity with a list of objects.
+func (c *Client) List(obj db.Entity) ([]db.Entity, error) {
+	pairs, _, err := c.client.KV().List(c.qualified(obj.Prefix()), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	entities := []db.Entity{}
+
+	for _, pair := range pairs {
+		copy, err := helpers.ReadAndSet(c, obj, pair.Key, pair.Value)
+		if err != nil {
+			return nil, err
+		}
+
+		entities = append(entities, copy)
+	}
+
+	return entities, nil
+}
+
+// ListPrefix lists all the entities under prefix instead of listing the whole keyspace.
+func (c *Client) ListPrefix(prefix string, obj db.Entity) ([]db.Entity, error) {
```

Comment:
`prefix` is passed in. `obj.Prefix()` is an interface method on `db.Entity` explained through `db/db.go`.